src/lib/packager/clients/pkg_finish.rb

```diff
@@ -0,0 +1,192 @@
+        ::FileUtils.cp(FAILED_PKGS_PATH, File.join(Installation.destdir, FAILED_PKGS_PATH),
+          preserve: true)
+      end
+    end
+
+  private
+
+    # Backup old sources
+    #
+    # During upgrade, old sources are reinitialized
+    # either as enabled or disabled.
+    # The old sources from target should go away.
+    def backup_all_target_sources
+      if !File.exist?(REPOS_DIR)
+        log.error("Directory #{REPOS_DIR} doesn't exist!")
+        return
+      end
+
+      current_repos = SCR.Read(path(".target.dir"), REPOS_DIR)
+
+      if current_repos.nil? || current_repos.empty?
+        log.warn("There are currently no repos in #{REPOS_DIR} conf dir")
+        return
+      else
+        log.info("These repos currently exist on a target: #{current_repos}")
+      end
+
+      # Backup repos.d
+      repos_backup_dir = File.join(Directory.vardir, "repos.d_backup")
+      backup_old_sources(REPOS_DIR, repos_backup_dir).nil?
+
+      # Clean old repositories
+      remove_old_sources(current_repos)
+
+      # Sync sources
+      sync_target_sources
+
+      nil
+    end
+
+    # Disable given repositories if needed
+    #
+    # Given a local repository:
+    #
+    # * if all products it contains are available through another repository,
+    #   then it will be disabled;
+    # * if some product contained is not available through another
+    #   repository, then it will be left untouched.
+    #
+    # @return [Array<Packages::Repository>] List of disabled repositories
+    def disable_local_repos
+      candidates_repos, other_repos = *::Packages::Repository.enabled.partition(&:local?)
+      products = other_repos.map(&:products).flatten.uniq
+      candidates_repos.each_with_object([]) do |repo, disabled|
+        uncovered = repo.products.reject { |p| products.include?(p) }
+        if uncovered.empty?
+          log.info("Repo #{repo.repo_id} will be disabled because products are present "\
+            "in other repositories")
+          repo.disable!
+          disabled << repo
+        else
+          log.info("Repo #{repo.repo_id} cannot be disabled because these products " \
+                   "are not available through other repos: #{uncovered.map(&:name)}")
+        end
+      end
+    end
+
+    # Backup sources
+    #
+    # @param source [String] Path of sources to backup
+    # @param target [String] Directory to store backup
+    # @return [String] Name of the backup archive (locate in the given target directory)
+    def backup_old_sources(source, target)
+      archive_name = "repos_#{Time.now.strftime(BACKUP_TIMESTAMP_FORMAT)}.tgz"
+      compress_cmd = format(TAR_CMD,
+        target: String.Quote(target),
+        archive: archive_name,
+        source: String.Quote(source))
+      cmd = SCR.Execute(path(".target.bash_output"), compress_cmd)
+      if !cmd["exit"].zero?
+        log.error("Unable to backup current repos; Command >#{compress_cmd}< returned: #{cmd}")
```

Comment:
If it fails then the returned value is the result of `log.error` which is wrong. It should either throw an exception or return `nil` on error.